r/tests/testthat/test-dplyr-funcs-datetime.R

```diff
@@ -2309,3 +2308,535 @@ test_that("build_formats() and build_format_from_order()", {
+          cob_null = ceiling_date(date, unit, change_on_boundary = NULL),
+          cob_true = ceiling_date(date, unit, change_on_boundary = TRUE),
+          cob_false = ceiling_date(date, unit, change_on_boundary = FALSE)
+        ) %>%
+        collect(),
+      boundary_times[r, ] %>%
+        mutate(
+          cob_null = as.Date(ceiling_date(date, unit, change_on_boundary = NULL)),
+          cob_true = as.Date(ceiling_date(date, unit, change_on_boundary = TRUE)),
+          cob_false = as.Date(ceiling_date(date, unit, change_on_boundary = FALSE))
+        ),
+      ...
+    )
+  }
+
+
+}
+
+test_that("ceiling_date() applies change_on_boundary correctly", {
+  check_boundary_with_unit(".001 second")
+  check_boundary_with_unit("second")
+  check_boundary_with_unit("minute", tolerance = .001) # floating point issue?
+  check_boundary_with_unit("hour")
+  check_boundary_with_unit("day")
+})
+
+
+# In lubridate, an error is thrown when 60 sec/60 min/24 hour thresholds are
+# exceeded. Checks that arrow mimics this behaviour and throws an identically
+# worded error message
+test_that("temporal round/floor/ceil period unit maxima are enforced", {
+
+  expect_error(
+    call_binding("round_date", Expression$scalar(Sys.time()), "61 seconds"),
+    "Rounding with second > 60 is not supported"
+  )
+  expect_error(
+    call_binding("round_date", Expression$scalar(Sys.time()), "61 minutes"),
+    "Rounding with minute > 60 is not supported"
+  )
+  expect_error(
+    call_binding("round_date", Expression$scalar(Sys.time()), "25 hours"),
+    "Rounding with hour > 24 is not supported"
+  )
+  expect_error(
+    call_binding("round_date", Expression$scalar(Sys.Date()), "25 hours"),
+    "Rounding with hour > 24 is not supported"
+  )
+
+})
+
+check_timezone_rounding <- function(data, unit) {
+  compare_dplyr_binding(
+    .input %>%
+      mutate(
+        utc_floored = floor_date(utc_time, unit = unit),
+        utc_rounded = round_date(utc_time, unit = unit),
+        utc_ceiling = ceiling_date(utc_time, unit = unit),
+        syd_floored = floor_date(syd_time, unit = unit),
+        syd_rounded = round_date(syd_time, unit = unit),
+        syd_ceiling = ceiling_date(syd_time, unit = unit),
+        adl_floored = floor_date(adl_time, unit = unit),
+        adl_rounded = round_date(adl_time, unit = unit),
+        adl_ceiling = ceiling_date(adl_time, unit = unit),
+        mar_floored = floor_date(mar_time, unit = unit),
+        mar_rounded = round_date(mar_time, unit = unit),
+        mar_ceiling = ceiling_date(mar_time, unit = unit),
+        kat_floored = floor_date(kat_time, unit = unit),
+        kat_rounded = round_date(kat_time, unit = unit),
+        kat_ceiling = ceiling_date(kat_time, unit = unit)
+      ) %>%
+      collect(),
+    data
+  )
+}
+
+test_that("timestamp rounding takes place in local time", {
+
+  # lubridate does not always return the correct results
+  # for some of our test cases in tz_times, esp on windows
```

Comment:
How bad is this? Is there a chance it's a bug in our C++ kernel?

Reply:
It's not us: it's either lubridate or something lower-level that lubridate relies on. The bug only occurs in very specific case where sub-second timing is relevant to the result of the rounding, only on windows, it's specific to zoned times (but not all timezones!) and it's lubridate that returns inconsistent answers, not arrow. As a paranoia check, I've added a second batch of tests that check that the arrow results give the same answer in local time for the zoned times as it does for the UTC times. I think we're in the clear

Reply:
Thank you for checking! This will help me sleep better.

It's odd that the problem occurs only for sub-seconds. Maybe data is stored as float at some point? Either way I think this is resolved!